eip-0041.md

```diff
@@ -0,0 +1,139 @@
+    // Name: EIP-0041 Stealth address contract
+    // Version: 1.0.0
+    val gr = SELF.R4[GroupElement].get
+    val gy = SELF.R5[GroupElement].get
+    val ur = SELF.R6[GroupElement].get
+    val uy = SELF.R7[GroupElement].get
+
+    proveDHTuple(gr,gy,ur,uy)
+}
+```
+
+**Script ErgoTree:**
+
+```
+1000cee4c6a70407e4c6a70507e4c6a70607e4c6a70707
+```
+
+**Generation of stealth box registers [[5]](#5):**
+
+```typescript
+const g = new EC("secp256k1").g; // group element generator
+const u = receiverPublicKey;
+const r = BigInt(rand(32));
+const y = BigInt(rand(32));
+const gr = g.mul(r); // gr = g^r = R4
+const gy = g.mul(y); // gy = g^y = R5
+const ur = u.mul(r); // ur = u^r = R6
+const uy = u.mul(y); // uy = u^y = R7
+```
+
+**Box register declarations:**
+
+- Register `R4`
+  - Type: `SConstant[SGroupElement]`
+  - Value: gr = g^r
+- Register `R5`
+  - Type: `SConstant[SGroupElement]`
+  - Value: gy = g^y
+- Register `R6`
+  - Type: `SConstant[SGroupElement]`
+  - Value: ur = u^r
+- Register `R7`
+  - Type: `SConstant[SGroupElement]`
+  - Value: uy = u^y
+
+> 📝 As discussed in the `ergoforum` discussion [[1]](#1) this register declaration is larger in size than that originally proposed by `kushti` but possesses the useful property that it could look like a legitimate use-case.
+
+## Public key generation
+
+Public keys used for `Receiver` stealth addresses should be generated following `EIP-3` [[6]](#6) so they are deterministic and payments made to `Receiver` can be retrieved after a fresh wallet restore.
+
+## Scanning for spendable stealth boxes
+
+All stealth boxes created following this standard will be located at the encoded address `2pxnLshTr2aL5t3QFd6ScSEHC5Jk6EpnMRzGb3`. In order to discover spendable stealth boxes a wallet would need to check the following for each box at this address:
+
+```ts
+const gr = stealthBox.R4;
+const gy = stealthBox.R5;
+const ur = stealthBox.R6;
+const uy = stealthBox.R7;
+const x = receiverSecretKey; // the secret key corresponding to the public key shared by `Receiver`
+const isSpendable = ur.eq(gr.mul(x)) && uy.eq(gy.mul(x));
+```
+
+> ⚠️ All stealth boxes being located at `2pxnLshTr2aL5t3QFd6ScSEHC5Jk6EpnMRzGb3` is a potential scalability issue related to this EIP. Wallets/indexing services will need to ensure stealth box scanning implementations are efficient.
+
+## Stealth address encoding
+
+To provide users of Ergo a seamless experience when interacting with stealth addresses there needs to be a way for wallets/applications to detect a `Receiver` is requesting a stealth payment. If a `Receiver` shares a standard `P2PK` address, for example, there is currently no way for a wallet to know that the user wants to receive a stealth payment.
+
+To address this issue we propose adding a new `AddressType`.
+
+Currently Ergo uses:
+
+```ts
+enum AddressType {
+    P2PK = 1,
+    P2SH = 2,
+    P2S = 3
+}
+```
+
+We propose adding a new address type of `AddressType.P2SA`:
```

Comment:
It would be good to describe how to encode a P2SA and make clear that it's a special kind of address, as it's not ErgoTree-based like others. 

You can take this comment as the basis for that: https://github.com/ergoplatform/eips/pull/87#discussion_r1051699979